flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -348,20 +353,42 @@ private void handleStartTaskExecutorServicesException(Exception e) throws Except
 		resourceManagerHeartbeatManager.stop();
 
 		try {
-			stopTaskExecutorServices();
+			stopServicesBeforeTasksCompletion();
 		} catch (Exception e) {
 			throwable = ExceptionUtils.firstOrSuppressed(e, throwable);
 		}
 
 		if (throwable != null) {
 			return FutureUtils.completedExceptionally(new FlinkException("Error while shutting the TaskExecutor down.", throwable));
 		} else {
-			log.info("Stopped TaskExecutor {}.", getAddress());
-			return CompletableFuture.completedFuture(null);
+			return FutureUtils.waitForAll(taskCompletionFutures).thenRun(() -> {
+				try {
+					stopServicesAfterTasksCompletion();
+				} catch (Exception e) {
+					throw new FlinkRuntimeException("Error while shutting the TaskExecutor down.", e);
+				}
+				log.info("Stopped TaskExecutor {}.", getAddress());
+			});
 		}
 	}
 
-	private void stopTaskExecutorServices() throws Exception {
+	private List<CompletableFuture<Void>> failTasks(JobManagerConnection jobManagerConnection,
+			Throwable cause) {
+		List<CompletableFuture<Void>> taskCompletionFutures = new ArrayList<>();
+		Iterator<Task> tasks = taskSlotTable.getTasks(jobManagerConnection.getJobID());
```

Comment:
Why not to use foreach loop instead of Iterator with while loop:
`for (Task task : taskSlotTable.getTasks(jobID))` (I would have jobID as `failTasks` argument)?

Reply:
Foreach loop is not applicable to Iterator but Iterable :)

Reply:
Oh, true, weird... `getTasks` could return `Iterable` but this would be a beforehand hotfix :)